### PR TITLE
support parsing sub-string

### PR DESCRIPTION
**Pull request by elahehrashedi**
Add support parsing define variables that are in the form of a string, e.g.:
`target_compile_definitions(prname PUBLIC AWESOME=\"foo\")`
bug:
main:
https://github.com/microsoft/vscode-cmake-tools/issues/969

duplicates: 
https://github.com/microsoft/vscode-cmake-tools/issues/1493
https://github.com/microsoft/vscode-cmake-tools/issues/1565

### CODE REVIEW

File: src/shlex.ts [hunk 1]
```diff
@@ -41,17 +42,29 @@ export function* split(str: string, opt?: ShlexOptions): Iterable<string> {
...
       }
-
       // Another quoted char
       token = (token || '') + char;
       continue;
     }
 
     if (quoteChars.includes(char)) {
+      // Beginning of a sub-quoted token
+      if (i > 1 && str.substring(i - 1, i + 1) === "\\\"") {
```
Comment (sean-mcmanus) [inline src/shlex.ts:61]: It seems like `"\\\'"` should be handled too, maybe just when when opt.mode == "posix".

**Review by sean-mcmanus — CHANGES_REQUESTED** [review]

**Review by sean-mcmanus — CHANGES_REQUESTED** [review]
The posix case can use ‘’ or “” quotes.

**Comment by bobbrow** [comment]
Can you add a test that covers this @elahehrashedi?

**Review by andreeis — APPROVED** [review]

**Review by sean-mcmanus — APPROVED** [review]

**Comment by sean-mcmanus** [comment]
> Can you add a test that covers this @elahehrashedi?

A test case appears to exist already at https://github.com/microsoft/vscode-cmake-tools/blob/80a745aeb7a12c76f5976a7b4ebca9674d17b5dd/test/unit-tests/shlex.test.ts, so it looks like you'd just need to add some lines to cover this scenario.

File: src/shlex.ts [hunk 2]
```diff
@@ -41,17 +42,29 @@ export function* split(str: string, opt?: ShlexOptions): Iterable<string> {
...
       }
-
       // Another quoted char
       token = (token || '') + char;
       continue;
     }
 
     if (quoteChars.includes(char)) {
+      // Beginning of a sub-quoted token
+      if (i > 0 && (opt.mode === 'posix' ? (str.substring(i - 1, i + 1) === "\\\'" || str.substring(i - 1, i + 1) === "\\\"") : str.substring(i - 1, i + 1) === "\\\"")) {
```
Comment (michelleangela) [inline src/shlex.ts:61]: `str.substring(i - 1, i + 1)` repeats 3 times, Can it be saved into a variable? And then compare that variable with the different `\\\`

**Review by michelleangela — CHANGES_REQUESTED** [review]

File: src/shlex.ts (hunk 2 above)
Comment (michelleangela) [reply src/shlex.ts:61]: `"\\\""` also repeats, Can the conditional be simplified to sometime like:

```
const quote: string = str.substring(i - 1, i + 1);
if (i > 0 && (quote === "\\\"" || ( opt.mode === 'posix' && === "\\\'"))

**Comment by elahehrashedi** [comment]
I am closing this PR, the content of this PR are included here: https://github.com/microsoft/vscode-cmake-tools/pull/1648

**Pull request closed without merging**